src/utils/applyChangeToValue.spec.js

```diff
@@ -0,0 +1,180 @@
+import expect from 'expect'
+import applyChangeToValue from './applyChangeToValue'
+import markupToRegex from './markupToRegex'
+
+describe('#applyChangeToValue', () => {
+  const userMarkup = '@[__display__](user:__id__)'
+  const emailMarkup = '@[__display__](email:__id__)'
+  const defaultDisplayTransform = (id, display) => display
+  const config = [
+    {
+      markup: userMarkup,
+      regex: markupToRegex(userMarkup),
+      displayTransform: defaultDisplayTransform,
+    },
+    {
+      markup: emailMarkup,
+      regex: markupToRegex(emailMarkup),
+      displayTransform: defaultDisplayTransform,
+    },
+  ]
+
+  const value =
+    "Hi @[John Doe](user:johndoe), \n\nlet's add @[joe@smoe.com](email:joe@smoe.com) to this conversation..."
+  const plainText =
+    "Hi John Doe, \n\nlet's add joe@smoe.com to this conversation..."
+
+  const displayTransform = id => `<--${id}-->`
+  const plainTextDisplayTransform =
+    "Hi <--johndoe-->, \n\nlet's add <--joe@smoe.com--> to this conversation..."
+
+  it('should correctly add a character at the end, beginning, and in the middle of text', () => {
+    let changed = 'S' + plainText
+    let result = applyChangeToValue(value, config, changed, 0, 0, 1)
```

Comment:
If possible I'd prefer a change the signature of the method. To me, everything that comes after `config` should just be inside config under a named key. I'll never be able to remember which number is which. Alternatively, there could be another `options` parameter which includes `plainTextValue` and the selection indices. However, this looks kinda weird as well.